examples/seismic/elastic/operators.py

```diff
@@ -3,38 +3,134 @@
+                      npoint=geometry.nrec)
     if model.grid.dim == 3:
-        src_yy = src.inject(field=tau[1, 1].forward, expr=src * s)
-        src_expr += src_yy
+        rec_vy = Receiver(name='rec_vy', grid=model.grid, time_range=geometry.time_axis,
+                          npoint=geometry.nrec)
+    name = "rec_tau" if forward else "rec"
+    rec = Receiver(name="%s" % name, grid=model.grid, time_range=geometry.time_axis,
+                   npoint=geometry.nrec)
 
-    # Create interpolation expression for receivers
-    rec_term1 = rec1.interpolate(expr=tau[-1, -1])
-    rec_term2 = rec2.interpolate(expr=div(v))
+    if forward:
 
-    return src_expr + rec_term1 + rec_term2
+        # The source injection term
+        src_xx = src.inject(field=tau[0, 0].forward, expr=src * s)
+        src_zz = src.inject(field=tau[-1, -1].forward, expr=src * s)
+        src_expr = src_xx + src_zz
+        if model.grid.dim == 3:
+            src_yy = src.inject(field=tau[1, 1].forward, expr=src * s)
+            src_expr += src_yy
+        # Create interpolation expression for receivers
+        rec_term_vx = rec_vx.interpolate(expr=v[0])
+        rec_term_vz = rec_vz.interpolate(expr=v[-1])
+        expr = tau[0, 0] + tau[-1, -1]
+        rec_expr = rec_term_vx + rec_term_vz
+        if model.grid.dim == 3:
+            expr += tau[1, 1]
+            rec_term_vy = rec_vy.interpolate(expr=v[1])
+            rec_expr += rec_term_vy
+        rec_term_tau = rec.interpolate(expr=expr)
+        rec_expr += rec_term_tau
+
+    else:
+        # Construct expression to inject receiver values
+        rec_xx = rec.inject(field=tau[0, 0].backward, expr=rec*s)
+        rec_zz = rec.inject(field=tau[-1, -1].backward, expr=rec*s)
+        rec_expr = rec_xx + rec_zz
+        expr = tau[0, 0] + tau[-1, -1]
+        if model.grid.dim == 3:
+            rec_expr += rec.inject(field=tau[1, 1].backward, expr=rec*s)
+            expr += tau[1, 1]
+        # Create interpolation expression for the adjoint-source
+        src_expr = src.interpolate(expr=expr)
+
+    return src_expr, rec_expr
+
+
+def elastic_stencil(model, v, tau, forward=True):
+
+    lam, mu, b, damp = model.lam, model.mu, model.b, model.damp
+
+    rho = 1. / b
+
+    if forward:
+
+        # Particle velocity
+        eq_v = v.dt - b * div(tau)
+        # Stress
+        e = (grad(v.forward) + grad(v.forward).T)
+        eq_tau = tau.dt - lam * diag(div(v.forward)) - mu * e
+
+        u_v = Eq(v.forward, damp * solve(eq_v, v.forward))
+        u_t = Eq(tau.forward, damp * solve(eq_tau, tau.forward))
+
+        return [u_v, u_t]
+
+    else:
+
+        """
+        Implementation of the elastic wave-equation from:
+        1 - Feng and Schuster (2017): Elastic least-squares reverse time migration
+        https://doi.org/10.1190/geo2016-0254.1
+
+        """
+
+        lpmu = lam + 2. * mu
+
+        pde_vx = rho * v[0].dt.T - (lpmu * tau[0, 0]).dx - (lam * tau[1, 1]).dx - \
```

Comment:
I'm sorry but I don't really like this. We have a nice Tensor forward PDE that looks very close to the math and clean and the adjoint has a Tensor expression as well this is not very readable and I would prefer a tensor adjoint equation as well.

Reply:
Thanks @mloubout for the review. 
I also totally agree that adjoint formulation is not very readable. But I couldn't find the vectorial/tensorial adjoint 
formulation appropriate for Devito.
Actually, I have derived a vectorial/tensorial adjoint equation but I need to calculate the `diag(grad(TensorTimeFunction))` and the Devito doesn't have this option. Because of that, I decided to use this kind of implementation.
Could you give some work or the math steps for finding the adjoint wave equation with a friendly Tensor expression for Devito?
Please see my vectorial/tensorial adjoint equation derivation attached.
[Elastic notes.pdf](https://github.com/devitocodes/devito/files/8919521/Elastic.notes.pdf)

Reply:
@mloubout 
I am investigating the possibilities of using a compact formulation for the elastic adjoint wave equation (equation 5 of this [paper](https://doi.org/10.1190/geo2016-0254.1)). I would like to know if you are sure that it is possible to write this equation using vectorial/tensorial formulation with the current features of Devito.
In my point of view, I suppose it is necessary to implement some features to do this, but in fact, I'm not sure.
So, from your answer, I decide if I keep the PR and persist trying to implement the vector/tensorial formulation or if I remove the PR and move   my efforts to implement the neccessary features that allow a vectorial/tensorial implementation of elastic adjoint equation.

Reply:
what would you expect `diag(grad(TensorTimeFunction))`to do ? you would be creating an N-dimensional tensor here, I'm not sure I follow

```
In [26]: from devito import diag

In [27]: print(diag.__doc__)

    Creates a diagonal tensor with func on its diagonal.

    Parameters
    ----------
    func : Differentiable or scalar
```

Reply:
ah wait... I've now looked at your notes. Did you actually mean `diag(div)` ? that's definitely supported, I wouldn't see why it shouldn't work

Reply:
is it an error in the maths ? you would be summing `diag(grad(TensorTimeFunction)` to `div(TensorTimeFunction)` which is a Vector, so that can't possibly be right